classes/services/PKPSubmissionService.php

```diff
@@ -0,0 +1,646 @@
+						break;
+
+					case 'submissionProgress':
+						$compiled[$param] = (int) $submission->getSubmissionProgress();
+						break;
+
+					case 'reviewRounds':
+						$compiled[$param] = $this->toArrayReviewRounds($submission);
+						break;
+
+					case 'reviewAssignments':
+						$compiled[$param] = $this->toArrayReviewAssignments($submission);
+						break;
+
+					case 'source':
+					case 'copyrightNotice':
+					case 'rights':
+						// @todo needs params
+						break;
+
+					case 'urlWorkflow':
+						$compiled[$param] = $this->getWorkflowUrlByUserRoles($submission);
+						break;
+
+					default:
+
+						$method = '';
+						if (method_exists($submission, 'getLocalized' . ucfirst($param))) {
+							$method = 'getLocalized' . ucfirst($param);
+						} elseif (method_exists($submission, 'get' . ucfirst($param))) {
+							$method = 'get' . ucfirst($param);
+						}
+						if (!empty($method)) {
+							$compiled[$param] = $submission->{$method}();
+						}
+						break;
+				}
+			}
+
+			$output[] = $compiled;
+		}
+
+		\HookRegistry::call('Submission::toArray::output', array(&$output, $params, $submissions));
+
+		return $output;
+	}
+
+	/**
+	 * Get details about a submission's stage(s)
+	 *
+	 * @param $submission Submission
+	 * @param $stageIds array|int|null One or more stages to retrieve.
+	 *  Default: null. Will return data on all app stages.
+	 * @return array {
+	 *  `id` int stage id
+	 *  `label` string translated stage name
+	 *  `queries` array [{
+	 *    `id` int query id
+	 *    `assocType` int
+	 *    `assocId` int
+	 *    `stageId` int
+	 *    `sequence` int
+	 *    `closed` bool
+	 *   }]
+	 *  `statusId` int stage status. note: on review stage, this refers to the
+	 *    status of the latest round.
+	 *  `status` string translated stage status name
+	 *  `files` array {
+	 *    `count` int number of files attached to stage. note: this only counts
+	 *      revision files.
+	 *   }
+	 */
+	public function toArrayStageDetails($submission, $stageIds = null) {
+
+		if (is_null($stageIds)) {
+			$stageIds = Application::getApplicationStages();
+		} elseif (is_int($stageIds)) {
+			$stageIds = array($stageIds);
+		}
+
+		$stages = array();
+		foreach ($stageIds as $stageId) {
+
+			import('lib.pkp.classes.workflow.WorkflowStageDAO');
+			$workflowStageDao = DAORegistry::getDAO('WorkflowStageDAO');
+			$stage = array(
+				'id' => (int) $stageId,
+				'label' => __($workflowStageDao->getTranslationKeyFromId($stageId)),
+				'isActiveStage' => $submission->getStageId() == $stageId,
+			);
+
+			// Discussions in this stage
+			$stage['queries'] = array();
+			$request = Application::getRequest();
+			import('lib.pkp.classes.query.QueryDAO');
+			$queryDao = DAORegistry::getDAO('QueryDAO');
+			$queries = $queryDao->getByAssoc(
+				ASSOC_TYPE_SUBMISSION,
+				$submission->getId(),
+				$stageId,
+				$request->getUser()->getId() // Current user restriction should prevent unauthorized access
+			)->toArray();
```

Comment:
The `toArray` function is relatively wasteful compared to just iterating through with `->next`. I don't think it's necessary to convert to an array here.